src/userdev/java/net/minecraftforge/gradle/userdev/util/Deobfuscator.java

```diff
@@ -0,0 +1,150 @@
+            SrgJarRenamer.rename(original, output, names);
+
+            Utils.updateHash(output, HashFunction.SHA1);
+            cache.save();
+        }
+
+        return output;
+    }
+
+    public File deobfSources(File original, String mappings, String... cachePath) throws IOException {
+        project.getLogger().debug("Deobfuscating sources file {} with mappings {}", original.getName(), mappings);
+
+        File names = findMapping(mappings);
+        if (names == null || !names.exists()) {
+            return null;
+        }
+
+        File cacheDir = getOrCreateCacheDir(cachePath);
+
+        String mappedName = getUniqueMappedName(original, mappings);
+
+        File input = new File(cacheDir, mappedName + ".input");
+        File output = new File(cacheDir, mappedName);
+
+        HashStore cache = new HashStore()
+                .load(input)
+                .add("names", names)
+                .add("orig", original);
+
+        if (!cache.isSame() || !output.exists()) {
+            McpNames map = McpNames.load(names);
+
+            try (ZipInputStream zin = new ZipInputStream(new FileInputStream(input));
+                 ZipOutputStream zout = new ZipOutputStream(new FileOutputStream(output))) {
+                ZipEntry _old;
+                while ((_old = zin.getNextEntry()) != null) {
+                    ZipEntry _new = new ZipEntry(_old.getName());
+                    _new.setTime(0);
+                    zout.putNextEntry(_new);
+
+                    if (_old.getName().endsWith(".java")) {
+                        String mapped = map.rename(zin, true);
+                        IOUtils.write(mapped, zout);
+                    } else {
+                        IOUtils.copy(zin, zout);
+                    }
+                }
+            }
+
+            Utils.updateHash(output, HashFunction.SHA1);
+            cache.save();
+        }
+
+        return output;
+    }
+
+    private File getOrCreateCacheDir(String... cachePath) {
+        File cacheDir = new File(cacheRoot, String.join(File.separator, cachePath));
+        ;
+
+        if (!cacheDir.exists()) {
+            cacheDir.mkdirs();
+        }
+
+        return cacheDir;
+    }
+
+    private File findMapping(String mapping) {
+        if (mapping == null)
+            return null;
+
+        int idx = mapping.lastIndexOf('_');
+        String channel = mapping.substring(0, idx);
+        String version = mapping.substring(idx + 1);
+        String desc = "de.oceanlabs.mcp:mcp_" + channel + ":" + version + "@zip";
+        return MavenArtifactDownloader.manual(project, desc, false);
+    }
+
+    private static String getUniqueMappedName(File file, String mappings) throws IOException {
+        String originalBaseName = file.getName().substring(0, file.getName().lastIndexOf('.'));
+        String originalExtension = file.getName().substring(originalBaseName.length());
+        String hash = HashFunction.SHA1.hash(file);
```

Comment:
Why is the hash here? And this derps up the whole classifier vs version thing. Where you'd end up with things like "foo/1.0/foo-1.0-sources_mapped_blah.jar" instead of what the remapping repo is expecting: "foo/1.0_mapped_blah/foo-1.0_mapped_blah-sources.jar"

Reply:
This is used for the cache, not for the artifacts downloaded through maven and since this class is the only one accessing them, it works fine. The version Repo gets is used to find the artifact and mappings version, so it's not affected by this in any way.

That said, this way, the cache doesn't follow the maven structure that way, but I didn't think much of it, since local files won't be following that in any case, and the cache is just internals. If that's an issue I could restore the way maven artifacts were stored before (so no hash for them) and have hashes only for local files. Should I?

Reply:
Use the maven format I have speced for the maven formats.
use a {ObjectHash}/filename_mapped_blah format for flat files. 

Tho HONESTLY I think we should nuke any concept of using a file based deobfed dep. As all mods SHOULD be releasing as uniquely identified maven artifacts. And thus we should encourage people to use that in dev as well. What setups do you think would be good to have a flat file system? Why is that better then a maven structure?

Reply:
> Use the maven format I have speced for the maven formats
Sure.

As for file deobf deps, it's basically an option for when no better ones are available. 

All mods should be releasing as maven artifacts, but unfortunately that's probably never gonna happen. So if a user wants to depend on someone else's mod that's released as a file (say on curseforge), but not as a maven artifact, they'd have to either try to get it through CF's maven, which is shaky at best (spaces in filenames just break it), or, if it's open-source, clone it, set up publishing and publish to local. And who knows what kind of shitty un-reproducible build that other mod has. In cases like that, it's way simpler to just download and use it as a file dependency.

Reply:
All files can be referenced as a maven artifact. It just needs to be in the correct folder structure.
As for curseforge, there already exists projects to turn CurseForge into a maven repo.

Reply:
So we don't handle local files and tell people that if they want to use one, they can add it with a `flatDir` repo? Not as convenient for users, but it does remove the only part of the PR that depends on gradle internal API.

Reply:
Yes. At the end of the day this *shouldn't* be a major use case. As modders *should* be publishing to public mavens. Hell Forge's maven is fairly public. We just need to add a signup system. And i'd rather get rid of internal API dependencies as much as possible.

Reply:
Fair. Removing file handling then.